### DISCUSSION TITLE
Bumpmaps and texture mapping - why does it repeat?

### DISCUSSION THREAD

**Discussion by lopsided** [mitsuba-renderer/mitsuba3#1131] [Q&A]
I have a polyhedral shape that I am fitting using some inverse rendering. This is working quite except the objects in real life have slight defects and imperfections on the surfaces. I thought I could model these with a bumpmap bsdf. However, when I try this it seems that the bumpmap is repeating on every surface of my shape. Maybe this is just the expected behaviour? Is there any way to get around this - a bumpmap for each surface for example? Or some other option I might have missed?

MWE:

```python
import matplotlib.pyplot as plt
import mitsuba as mi
import torch

mi.set_variant('llvm_ad_rgb')

from mitsuba import ScalarTransform4f as T


def plot_scene():
    """
    Show a scene of a cube with a bumpmap texture.
    """

    # Create a bumpmap texture with a single bump in the center.
    bumpmap = torch.zeros((100, 100))
    bumpmap[40:60, 40:60] = 1
    bumpmap = mi.Bitmap(mi.TensorXf(bumpmap))

    # Create scene
    scene = mi.load_dict({
        'type': 'scene',
        'integrator': {'type': 'path'},
        'sensor': {
            'type': 'perspective',
            'to_world': T.look_at(target=[0, 0, 0], origin=[0, 0, 5], up=[0, 1, 0]),
            'sampler': {'type': 'independent', 'sample_count': 256},
            'film': {'type': 'hdrfilm', 'width': 200, 'height': 200},
        },
        'light': {
            'type': 'point',
            'to_world': T.look_at(target=[0, 0, 0], origin=[0, 0, 5], up=[0, 1, 0]),
            'intensity': {'type': 'spectrum', 'value': 100.0}
        },
        'cube': {
            'type': 'cube',
            'to_world': T.rotate([1, 0, 0], 45) @ T.rotate([0, 1, 0], 45),
            'bsdf': {
                'type': 'bumpmap',
                'bsdf': {'type': 'diffuse'},
                'bump_texture': {
                    'type': 'bitmap',
                    'raw': True,
                    'bitmap': bumpmap,
                    'wrap_mode': 'clamp',
                },
            }
        }
    })

    # Render scene
    image = mi.render(scene)
    plt.imshow(image**(1.0 / 2.2))
    plt.axis('off')
    plt.show()


if __name__ == '__main__':
    plot_scene()
```

Output:
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/1258160/f6890bc0-6b1a-4e1c-834d-9c2af06eab2e)

Is there any way I can get the bump to appear on just one of the cube faces?

Thanks!

**Comment by njroussel** [comment]
Hi @lopsided 

This is a bit weird, but expected. The `cube` defines a UV mapping that is replicating the texture per face. In order to only have it apply to one face, you'll have to change the UV-mapping which you can access and modify as follows:
```python
params = mi.traverse(scene)
params['cube.vertex_texcoords'] = ...
```

Alternatively, you could build your own cube mesh with the correct UV-mapping and then loading it as a `ply` or `obj` plugin.

**Reply by lopsided** [reply]
Hi @njroussel thanks for the quick reply!

The parameter looks like this:
```python
params['cube.vertex_texcoords'].numpy()
array([0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0.])   # shape=(48,)
```
If I understand correctly these are the flattened `uv` coordinates for each vertex? With 4 triangular faces on each cube face (`6*4*2=48`). Or, 
```python
dr.unravel(mi.Point2f, params['cube.vertex_texcoords']).numpy().shape  # (24, 2)
```
So to make a bump on just one face I would first split the `[0,1]x[0,1]` space up into 6 areas and convert the 0s and 1s to the uv coordinates that correspond to those faces Eg, the first triangular face might be something like `[0,0],[1/6,1/6],[1/3,0]`. Then put the bump in the middle of the face on the uv plane, eg, at `[1/6,1/6]`? Makes sense so far...

For my real problem I'm creating a procedural mesh like this:
```python
mesh = mi.Mesh(
    'shape',
    vertex_count=nv,
    face_count=nf,
    has_vertex_normals=False,
    has_vertex_texcoords=False,
    props=props  # bsdf in here
)
mesh_params = mi.traverse(mesh)
mesh_params['vertex_positions'] = dr.ravel(vertices)
mesh_params['faces'] = dr.ravel(faces)
```
but here the `texcoords` is empty:
```python
params['shape.vertex_texcoords'].numpy()
array([])   # shape=(0,)
```
It seems to do the same repeating texture thing that the cube does. Is this just the default behaviour? If I want to set a texture to cover the whole object should I set the `texcoords` after creating the scene or is there a better way of setting it directly when creating the mesh?

**Reply by lopsided** [reply]
Whoops I just realised I was setting `has_vertex_texcoords=False` in the constructor! If I set this `=True` then I get a big list of 0s
